sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -200,4 +199,15 @@ object CollationTypeCasts extends TypeCoercionRule {
         }
     }
   }
+
+  /**
+   * Returns whether the expression has explicit collated output type.
+   */
+  def hasExplicitCollation(expr: Expression): Boolean = {
+    expr match {
+      case _: Collate => true
+      case alias: Alias => hasExplicitCollation(alias.child)
```

Comment:
Are we sure this is safe to do for all other expressions not related to your PR? It might be that some expressions gets a subquery as input, could you please make sure we do not break something with propagating explicit collation like this?

Reply:
I will add a test case for this with an expression that gets the subquery as its input in these set operators